_posts/2020-04-07-trying-out-julia.md

```diff
@@ -0,0 +1,227 @@
+
+## What to do
+
+As said I have some background with MatLab and wanted to do some visualizations as they seem to be the strong point of Julia. So what data to visualize, something current...surprise surprise let's use COVID-19 data. Basic status data can be acquired from [api.covid19api.com](https://api.covid19api.com/) using REST calls. So basically to get day to day confirmed cases from Finland issue GET call to [api.covid19api.com/country/finland/status/confirmed](https://api.covid19api.com/country/finland/status/confirmed) and you get an array of the following json objects
+
+```json
+{
+    "Country": "Finland",
+    "Province": "",
+    "Lat": 61.9241,
+    "Lon": 25.7482,
+    "Date": "2020-03-31T00:00:00Z",
+    "Cases": 1418,
+    "Status": "confirmed"
+},
+```
+
+After some document browsing I know I need to add some packages to my Julia environment 
+
+| Package    | Purpose                                           |
+| :--------- | ------------------------------------------------- |
+| HTTP       | To make HTTP calls                                |
+| JSON       | Json tools for e.g. parsing                       |
+| Plots      | Basic plotting support                            |
+| StatsPlots | Extended plotting support e.g. dataframes support |
+
+This is achieved with the package manager in REPL by first issuing the ]-mark and then just `"activate <env_name>"` and then `"add HTTP"` and so on. Quite handy, I especially like that the package manager is baked into the REPL and not a completely separate tool. Now for some basic coding
+
+```julia
+using HTTP, JSON, Plots, StatsPlots
+
+resp = HTTP.get("https://api.covid19api.com/country/finland/status/confirmed")
+str = String(resp.body)
+jobj = JSON.Parser.parse(str)
+
+```
+
+So what I do here is first issue the HTTP get call to fetch the data from the REST API and then read the body to a string from the stream. Json parsing is quite simple and it produces a dict object. Let's see what it looks like
+
+```julia
+println(jobj[0])
+>>>
+BoundsError: attempt to access 65-element Array{Any,1} at index [0]
+in top-level scope at finland_confirmed.jl:6
+in getindex at base/array.jl:787
+```
+
+Oh right, should have seen this coming, as with MatLab in Julia indexes start from 1!!!
+
+```julia
+println(jobj[1])
+>>>
+("Lon" => 0,"Status" => "confirmed","Lat" => 0,"Date" => "2020-01-29T00:00:00Z",
+"Cases" => 1,"Country" => "Finland","Province" => "")
+```
+
+So now I can easily use the map function to modify it to a format suited for me. 
+
+```julia
+cases = map(x -> get(x, "Cases", 0), jobj)
+plot(cases)
+
+savefig("images/cases_simple.png") #Save plot as png image
+```
+
+The map function takes a mapping function and the data to apply the function to. Here I pass a anonymous function to pick up the Cases field and the data I got from the API call. Now that the data is a nice array of values let's plot it and see what's what.
```

Comment:
Jättäisin tuon Here:n pois tuosta Here I pass kohdatasta.

 Now that the data is a nice array of values let's plot it and see what's what

Tässä on varmaankin kaksi lausetta.

Reply:
Hmm, miksi kaksi lausetta? En ihan saanut kiinni tuosta.